### PR TITLE
Enhance the UX for packages with License Expressions

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug
Fixes: https://github.com/NuGet/Home/issues/7391
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 
Add license expression details to the Details and License Acceptance page. 

nonstandardlicense
complexlicense
complexnonstandardlicense
requirelicenseacceptance

## Testing/Validation
Tests Added: Yes, where appropriate.  
Reason for not adding tests:  
Validation done:

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs [hunk 1]
```diff
@@ -19,6 +19,9 @@
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
 using Task = System.Threading.Tasks.Task;
+using NuGet.Packaging;
+using NuGet.Packaging.Licenses;
+using System.Collections.Specialized;
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs]: fix ordering

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        private static IList<IText> GenerateLicenseLinks(LicenseMetadata licenseMetadata, Uri licenseUrl)
+        {
+            IList<IText> list = new List<IText>();
+            if (licenseMetadata != null)
+            {
+                list = GenerateLicenseLinks(licenseMetadata);
+            }
+            else
+            {
+                if (licenseUrl != null)
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: combined into a `else if`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+            if (metadata.WarningsAndErrors != null)
+            {
+                list.Add(new WarningText(string.Join(Environment.NewLine, metadata.WarningsAndErrors)));
+            }
+
+            return list;
+        }
+
+        private static void GetLicenseIdentifiers(NuGetLicenseExpression expression, IList<string> identifiers)
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: either return the identifier list or change the name to populateLicenseIdentifier

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+                        foreach (var identifier in identifiers)
+                        {
+                            var licenseStart = licenseToBeProcessed.IndexOf(identifier);
+                            if (licenseStart != 0)
+                            {
+                                list.Add(new FreeText(licenseToBeProcessed.Substring(0, licenseStart)));
+                            }
+                            var license = licenseToBeProcessed.Substring(licenseStart, identifier.Length);
+                            list.Add(new LicenseText(license, new Uri($"https://spdx.org/licenses/{license}.html")));
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: create a constant for the spdx url

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                    }
+                    else
+                    {
+                        list.Add(new FreeText(metadata.License));
+                    }
+
+                    break;
+                case LicenseType.File:
+
+                    list.Add(new FreeText($"License '{metadata.License}' is embedded in the package."));
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: add a resource string and do we need to localize this?

File: src/NuGet.Clients/NuGet.Tools/NuGetUIFactory.cs [hunk 6]
```diff
@@ -61,7 +61,8 @@ internal sealed class NuGetUIFactory : INuGetUIFactory
         public NuGetUIFactory(
             ICommonOperations commonOperations,
             INuGetUILogger logger,
-            ISourceControlManagerProvider sourceControlManagerProvider)
+            ISourceControlManagerProvider sourceControlManagerProvider,
```
Comment (jainaashish) [inline src/NuGet.Clients/NuGet.Tools/NuGetUIFactory.cs]: skip this from your PR. Pato will take care of that

File: src/NuGet.Core/NuGet.Packaging/NuspecReader.cs [hunk 7]
```diff
@@ -471,7 +471,29 @@ public LicenseMetadata GetLicenseMetadata()
...
+                                        if (!nugetLicense.IsStandardLicense)
+                                        {
+                                            if (invalidLicenseIdentifiers == null)
+                                            {
+                                                invalidLicenseIdentifiers = new List<string>();
+                                            }
+                                            invalidLicenseIdentifiers.Add(nugetLicense.Identifier);
+                                        }
+                                    };
+                                    expression.OnEachLeafNode(licenseProcessor, null);
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Packaging/NuspecReader.cs]: its so hard to read what exactly you're trying to do here...

File: test/TestUtilities/Test.Utility/JsonData.cs [hunk 8]
```diff
@@ -4651,5 +4651,111 @@ public static class JsonData
...
+          ""commitTimeStamp"": ""2015-02-02T23:04:00.0649617Z"",
+          ""catalogEntry"": {{
+            ""@id"": ""https://api.nuget.org/v3/catalog0/data/2015.02.02.19.38.08/PackageA.1.0.0.json"",
+            ""@type"": ""PackageDetails"",
+            ""authors"": ""PackageA startup components contributors"",
+            ""description"": ""PackageA IAppBuilder startup interface"",
+            ""iconUrl"": """",
+            ""id"": ""PackageA"",
+            ""language"": """",
+            ""licenseUrl"": ""https://github.com/PackageA-contrib/PackageA-hosting/blob/master/LICENSE.txt"",
```
Comment (jainaashish) [inline test/TestUtilities/Test.Utility/JsonData.cs:4687]: Would server also return license expression in search metadata for newer clients? or we always going to read it from nuspec file explicitly to show in browse tab?

**Review by jainaashish — APPROVED** [review]
Few small comments, overall looks good.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: Changed it to populate.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: I was gonna change this before merging. 
It'll be our Url, I will create a constant for it regardless though :)

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: Added

File: src/NuGet.Clients/NuGet.Tools/NuGetUIFactory.cs (hunk 6 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Tools/NuGetUIFactory.cs]: I have merged Pato's PR, will rebase from dev.

File: src/NuGet.Core/NuGet.Packaging/NuspecReader.cs (hunk 7 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/NuspecReader.cs]: I will add some comments. 
But basically this is a visitor for each leaf node, and if it sees any non standard licenses it creates a warning message.

File: test/TestUtilities/Test.Utility/JsonData.cs (hunk 8 above)
Comment (nkolev92) [reply test/TestUtilities/Test.Utility/JsonData.cs:4687]: Yes, the server will return a licenseExpression as per the design of @agr.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+namespace NuGet.PackageManagement.UI
+{
+    public interface IText
+    {
+        string Text { get; }
+    }
+
+    internal class LicenseText : IText
+    {
+        public LicenseText(string text, Uri link)
```
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs]: can `text` and `link` be null? If not, add null checks

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+using System;
+
+namespace NuGet.PackageManagement.UI
+{
+    public interface IText
+    {
+        string Text { get; }
+    }
+
+    internal class LicenseText : IText
```
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs]: I would suggest to have `LicenseText`, `FreeText` and `WarningText` in their own file for better discovery of the code. I don't have an strong opinion, so if you feel strongly against it disregard this comment.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+        public string Text { get; set; }
+        public Uri Link { get; set; }
+    }
+
+    internal class FreeText : IText
+    {
+        public FreeText(string text)
+        {
+            Text = text;
```
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs]: add null check if needed

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        public string Text { get; }
+    }
+
+    internal class WarningText : IText
+    {
+        public WarningText(string text)
+        {
+            Text = text;
```
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs]: add null check if needed

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        {
+            var list = new List<IText>();
+            switch (metadata.Type)
+            {
+                case LicenseType.Expression:
+
+                    if (metadata.LicenseExpression != null)
+                    {
+                        IList<string> identifiers = new List<string>();
+                        PopulateLicenseIdentifiers(metadata.LicenseExpression, ref identifiers);
```
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: if you are only adding you don't need to have the param as `ref`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+        internal static IReadOnlyList<IText> GenerateLicenseLinks(LicenseMetadata metadata)
+        {
+            var list = new List<IText>();
+            switch (metadata.Type)
+            {
+                case LicenseType.Expression:
+
+                    if (metadata.LicenseExpression != null)
+                    {
+                        IList<string> identifiers = new List<string>();
```
Comment (PatoBeltran) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: nit: `var identifiers`

File: src/NuGet.Core/NuGet.Packaging/NuspecReader.cs [hunk 15]
```diff
@@ -432,7 +432,7 @@ public LicenseMetadata GetLicenseMetadata()
 
                 var isKnownType = Enum.TryParse(type, ignoreCase: true, result: out LicenseType licenseType);
 
-                IList<string> errors = null;
+                List<string> errors = null;
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/NuspecReader.cs:435]: why change this from `IList` to `List`?

**Review by PatoBeltran — COMMENTED** [review]
I think the warning icon should be left of the text. This way if its consistent with most icons in VS and also if the text is really big, and the icon is right of the text it might get hidden and make the user need to scroll to see it.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs (hunk 10 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/IText.cs]: I'll move them to separate files.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs (hunk 13 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/PackageLicenseUtilities.cs]: good catch.
I think initially I was assigning that reference inside.

File: src/NuGet.Core/NuGet.Packaging/NuspecReader.cs (hunk 15 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/NuspecReader.cs:435]: I changed the warnings and errors to IReadOnlyList from IList. 

And List implements both :)

**Comment by nkolev92** [comment]
@PatoBeltran 
Addressed your feedback. 

Moved the warning in front of the text.
warningsignfirst

Can you peek at it again.

**Comment by nkolev92** [comment]
All the tests failing are due to the repo signing.

**Review by PatoBeltran — APPROVED** [review]

**Pull request merged**